### ISSUE TITLE
Need a way to sign GHfVS when script directory doesn't exist

### ISSUE THREAD

**Issue by jcansdale** [github/VisualStudio#946]
- GitHub Extension for Visual Studio  version: 2.2.0.9+
- Visual Studio version: All

These links might help:
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/public-signing.md
https://github.com/dsplaisted/strongnamer

**Comment by jcansdale** [comment]
I think the way to proceed is to publicly sign our assemblies when the script directory doesn't exist.

The public key for any assembly can be extracted like this:
```
sn -e GitHub.VisualStudio.dll Key.public
```

It can then be used to publicly sign an assembly. 

The .csproj would look like this:
```
<PropertyGroup>
    <AssemblyOriginatorKeyFile>..\..\Key.public</AssemblyOriginatorKeyFile>
    <SignAssembly>true</SignAssembly>
    <PublicSign>true</PublicSign>
    <DelaySign>false</DelaySign>
  </PropertyGroup>
```

**Comment by jcansdale** [comment]
See #965 for reasons why we sign `GitHub.*` assemblies *and* their dependencies with a unique key:

> @shana - We can certainly put an alternate public signing key in the repo that non-official builds can use, because signing is still a part of versioning and that way people don't have to do that manual part on their own (which they can).

#928 is a mitigation for people breaking us with alternative versions of our assemblies or - just as dangerously - distributing our signed assemblies with their extension. This is easy to do accidentally with VSIX projects (in fact developers need to go out of their way to not include referenced assemblies). Since the first loaded assembly gets to define the codebase, assemblies are easily orphaned from their dependencies! #923 is an example of Visual Studio breaking us in this way.

#928 sweeps up all previous versions of an assembly, so developers won't be tempted to work around this issue by referencing old versions of the `GitHub.*` assemblies.

**Comment by shana** [comment]
It's not only GitHub.* assemblies we have to worry about, we're not providing binding redirections for all our dependencies that we build and ship, and those are the ones that are very likely to get us blown up, so #928 is not a full solution to the versioning breaking problem here, as far as I can see.

**Comment by jcansdale** [comment]
@shana You're right #928, it's only a partial mitigation some relatively likely scenarios. 

For example someone could reference `GitHub.Exports` and have their extension work sometimes (depending on load order). Other times it won't work and they break us. 😢 @grokys ran into exactly this when he started working on inline comments.

It's only a partial solution, but it's better than nothing. 😄

It currently protects the following assemblies:
```
GitHub.UI
GitHub.VisualStudio.UI
GitHub.Exports
GitHub.Extensions
Octokit
LibGit2Sharp
Splat
```

I've tried a couple of times to provide a codebase for the rest, but for some strange reason it didn't work. 😕

**Comment by shana** [comment]
The more protected we are, the better! I don't think we should ever put our key in the public repo or encourage people to use it for signing, to ensure that we know what we always know what we're running. We can put a public key in the repo for signing when the internal one isn't available so it's easy for people to do - they still need to provide a client app ID and secret, so the whole thing is not completely automatic and should never be, so that users don't accidentally shoot themselves in the foot.

We're not a library for other people to use, we're an application. Users wanting to fork and tinker with it should have manual steps to do so that they are aware of what they're running and what they're doing, not just blindly hit a button and have things happen. Public mozilla code changes the icons and branding, for instance, so that it's clear that what you're building and running isn't the same thing that the system has. Small obstacles are good for awareness sometimes 😄
